airflow/www_rbac/decorators.py

```diff
@@ -125,3 +125,46 @@ def wrapper(self, *args, **kwargs):
                                         __class__.__name__ + ".login"))
         return wrapper
     return decorator
+
+def has_dag_access_ti(**dag_kwargs):
+    """
+    Decorator to check whether the user has read / write permission on the dag.
+    """
+    def decorator(f):
+        @functools.wraps(f)
+        def wrapper(self, *args, **kwargs):
+            has_access = self.appbuilder.sm.has_access
+            tis = args[0]
+            # if it is false, we need to check whether user has write access on the dag
+            can_dag_edit = dag_kwargs.get('can_dag_edit', False)
+            dag_ids = []
+            for ti in tis:
+                dag_ids.append(ti.dag_id)
+
+            # 1. check whether the user has can_dag_edit permissions on all_dags
+            # 2. if 1 false, check whether the user
+            #    has can_dag_edit permissions on the dag
+            # 3. if 2 false, check whether it is can_dag_read view,
+            #    and whether user has the permissions
+
+            if has_access('can_dag_edit', 'all_dags') or (not can_dag_edit and
+                  (has_access('can_dag_read','all_dags'))):
+                return f(self, *args, **kwargs)
+
+            failed_on = None
+
+            for dag_id in dag_ids:
+                if not (
```

Comment:
Can you explain a bit why its not:
```
if not (has_access('can_dag_edit', dag_id) or can_dag_edit)
```
?

Reply:
If we get to this point, then we dont have either can_dag_edit or can_dag_read on all dags.

So at this point, the user needs the right permission on the dag level
So you either need can_dag_edit on dag_id. If you dont have that and the parameter is can_dag_edit is False, then you need can_dag_read on dag_id. If either of those permissions are not true, then we should error out here.

Reply:
I think if a user only has `can_dag_read` on that dag_id, that should only give them the permission to read but not edit the dag (hence setting the task instance state)

Reply:
so I think just checking `has_access('can_dag_edit', dag_id)` should be sufficient.